### PR TITLE
Allow `FieldTimeSeries` to pass keyword arguments to `jldopen`

### LINKED ISSUE #3680: Support passing keyword arguments to `jldopen` when using `FieldTimeSeries`
I am interested in reading JLD2 files in parallel using multiple threads. This is actually now supported by JLD2.jl after https://github.com/JuliaIO/JLD2.jl/pull/477 provided you pass the `parallel_read = true` kwarg to `jldopen`.

I'm hoping to do this with `FieldTimeSeries`, but right now there's no way to pass kwargs to `jldopen` like there is with `JLD2OutputWriter`. I'm hoping to open a PR that adds support for this, but I'm just opening this issue in case there's any opposition or discussion to be had.

### PR DESCRIPTION
**Pull request by ali-ramadhan**
This PR takes some code I had to pass kwargs to `jldopen` through `FieldTimeSeries`.

But from the dicussion in issue #3680 it sounds like maybe a more future-proof way to achieve this is to add the kwargs object to the backend object (either `InMemory` or `OnDisk`)?

I guess there's a bit of a conflict of wording here? Backend refers to how the data is stored for a `FieldTimeSeries`. So we could end up with `OnDisk{NetCDF}` and `OnDisk{JLD2}` for example. And maybe also for `PartlyInMemory`. But for fully `InMemory` there's no "backend"?

Resolves #3680

### CODE REVIEW

**Comment by glwagner** [comment]
Right `InMemory` is more like a `nothing` backend, there's nothing back there. 

On ClimaOcean there's an example how we extended this infrastructure to a particular kind of NetCDF backend: https://github.com/CliMA/ClimaOcean.jl/blob/dd9148a8f702699ddf2947d3a1a094adddbd9bb1/src/DataWrangling/JRA55.jl#L218

The implementation in this PR seems simple enough though. @ali-ramadhan maybe can you provide a little syntax example how you expect this to be used? @navidcy take a look and see what you think

**Comment by ali-ramadhan** [comment]
I've used it to open JLD2 files in parallel using multiple threads with `@threads` so that each thread can process one time snapshot. Here's an example:

```julia
u = FieldTimeSeries("fields.jld2", "u";
    backend = OnDisk(),
    backend_kw = Dict(:parallel_read => true)
)

Nt = length(u)
temporal_reduction = zeros(Nt)

Threads.@threads for n in 1:Nt
    temporal_reduction[n] = median(fts[n])
end
```

Of course, this post-processing is much slower than just doing it on-the-fly as the simulation runs. But it's still useful to be able to do this.

One especially useful use case is parallel plotting from a `FieldTimeSeries`.

**Comment by glwagner** [comment]
Nice!

Only API suggestion is that the extra kw could be directly passed to `FieldTimeSeries`, so

```julia
u = FieldTimeSeries("fields.jld2", "u"; backend = OnDisk(), parallel_read=true)
```

Under the hood this can work with something like

```julia
function FieldTimeSeries(path::String, name::String; backend=InMemory(), kw...)
    # etc
    backend_kw = Dict(kw)
    # etc
end
```

You're the primary user though so you can state your preference!

**Comment by ali-ramadhan** [comment]
I agree that `kw...` results in a cleaner API with no passing of dicts.

I'm just wondering if there will be a use case in the future that will require passing a different set of kwargs. Then we'll need `kw1, kw1` as `kw...` cannot disambiguate between `kw1` and `kw2`.

**Comment by ali-ramadhan** [comment]
I thought about this a bit more and further think we should stick to keeping `kw` a dict so that the API is more consistent with how the `JLD2OutputWriter` constructor works as it also expects a dict.

**Comment by glwagner** [comment]
> I thought about this a bit more and further think we should stick to keeping `kw` a dict so that the API is more consistent with how the `JLD2OutputWriter` constructor works as it also expects a dict.

We could also change `JLD2OutputWriter`!

**Comment by ali-ramadhan** [comment]
Definitely agree the two interfaces should be consistent. I'd be okay with either, although still slightly favor the current interface in case we will want to pass two sets of kwargs in the future.

Would you support merging this PR and we can open an issue to discuss the best interface?

**Comment by glwagner** [comment]
Yes for sure support that

**Comment by glwagner** [comment]
Was hoping @navidcy might review but I can give it one

File: src/OutputReaders/field_time_series.jl [hunk 1]
```diff
@@ -380,8 +384,8 @@ function FieldTimeSeries(loc, grid, times=();
         isnothing(name) && error(ArgumentError("Must provide the keyword argument `name` when `backend=OnDisk()`."))
     end
     
-    return FieldTimeSeries{LX, LY, LZ}(data, grid, backend, boundary_conditions,
-                                       indices, times, path, name, time_indexing)
+    return FieldTimeSeries{LX, LY, LZ}(data, grid, backend, boundary_conditions, indices,
+                                       times, path, name, time_indexing, backend_kw)
```
[the remark this answers is not in the document]
Comment (ali-ramadhan) [reply src/OutputReaders/field_time_series.jl:388]: Ah true. I guess originally we didn't envision passing `FieldTimeSeries` into GPU kernels and I don't see a test that does this.

But with features like `FieldTimeSeries` forcing (PR #3760) an `adapt_structure` will be necessary. Actually not sure how tests will pass in that PR without adapting.

I can add some tests here though. There are no tests that use `backend_kw`...

[the remark this answers is not in the document]
Comment (glwagner) [reply src/OutputReaders/field_time_series.jl:388]: Ok but this PR breaks `FieldTimeSeries` on GPU which will break everything in ClimaOcean. @simone-silvestri can you add some tests here?

**Comment by ali-ramadhan** [comment]
Thanks @simone-silvestri for adding a GPU test for `FieldTimeSeries`! I've adapted it correctly now and tests pass locally.

**Comment by ali-ramadhan** [comment]
Updated the branch and added tests that `backend_kw` works with `FieldTimeSeries` and `FieldDataset` so I think this PR is actually ready for review now!

**Comment by ali-ramadhan** [comment]
Just bumping my review request :pleading_face:

File: test/test_output_readers.jl [hunk 2]
```diff
@@ -371,6 +371,36 @@ end
...
+            @test b3[1] isa Field
+        end
+    end
+
+    for Backend in [InMemory, OnDisk]
+        @testset "FieldDataset{$Backend} parallel reading" begin
+            @info "  Testing FieldDataset{$Backend} parallel reading..."
+
+            backend_kw = Dict(:parallel_read => true)
+            ds = FieldDataset(filepath3d, backend=Backend())
```
Comment (simone-silvestri) [inline test/test_output_readers.jl:394]:
```suggestion
            ds = FieldDataset(filepath3d; backend=Backend(), backend_kw)
```

**Review by simone-silvestri — APPROVED** [review]

**Comment by ali-ramadhan** [comment]
Curious if @glwagner has any thoughts on `backend_kw` vs. `reader_kw`, but I'm thinking of renaming to `reader_kw` then potentially revising the naming in a future refactor of `OutputReaders` (probably as part of supporting NetCDF).

**Pull request merged**